pkg/tfbridge/detailed_diff_test.go

```diff
@@ -1747,3 +1752,573 @@ func TestDetailedDiffPulumiSchemaOverride(t *testing.T) {
+			),
+			resource.NewPropertyMapFromMap(
+				map[string]interface{}{
+					"foo": computedValue,
+				},
+			),
+			tfs, ps,
+			map[string]*pulumirpc.PropertyDiff{
+				"foo": {Kind: pulumirpc.PropertyDiff_ADD},
+			},
+		)
+	})
+
+	t.Run("empty to computed", func(t *testing.T) {
+		runDetailedDiffTest(t,
+			propertyMapElems(),
+			resource.NewPropertyMapFromMap(
+				map[string]interface{}{
+					"foo": computedValue,
+				},
+			),
+			tfs, ps,
+			map[string]*pulumirpc.PropertyDiff{
+				"foo": {Kind: pulumirpc.PropertyDiff_UPDATE},
+			},
+		)
+	})
+
+	t.Run("two added, two removed", func(t *testing.T) {
+		runDetailedDiffTest(t,
+			propertyMapElems("val1", "val2"),
+			propertyMapElems("val3", "val4"),
+			tfs, ps,
+			map[string]*pulumirpc.PropertyDiff{
+				"foo[0]": {Kind: pulumirpc.PropertyDiff_UPDATE},
+				"foo[1]": {Kind: pulumirpc.PropertyDiff_UPDATE},
+			},
+		)
+	})
+
+	t.Run("two added, two removed, shuffled", func(t *testing.T) {
+		runDetailedDiffTest(t,
+			propertyMapElems("stable1", "stable2", "val1", "val2"),
+			propertyMapElems("val4", "val3", "stable1", "stable2"),
+			tfs, ps,
+			map[string]*pulumirpc.PropertyDiff{
+				"foo[0]": {Kind: pulumirpc.PropertyDiff_ADD},
+				"foo[1]": {Kind: pulumirpc.PropertyDiff_ADD},
+				"foo[2]": {Kind: pulumirpc.PropertyDiff_DELETE},
+				"foo[3]": {Kind: pulumirpc.PropertyDiff_DELETE},
+			},
+		)
+	})
+}
+
+func TestDetailedDiffSetBlock(t *testing.T) {
+	propertyMapElems := func(elems ...string) resource.PropertyMap {
+		var elemMaps []map[string]interface{}
+		for _, elem := range elems {
+			elemMaps = append(elemMaps, map[string]interface{}{"bar": elem})
+		}
+		return resource.NewPropertyMapFromMap(
+			map[string]interface{}{
+				"foo": elemMaps,
+			},
+		)
+	}
+
+	for _, forceNew := range []bool{false, true} {
+		sdkv2Schema := map[string]*schema.Schema{
+			"foo": {
+				Type: schema.TypeSet,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"bar": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: forceNew,
+						},
+					},
+				},
+			},
+		}
+		ps, tfs := map[string]*info.Schema{}, shimv2.NewSchemaMap(sdkv2Schema)
+		t.Run(fmt.Sprintf("forceNew=%v", forceNew), func(t *testing.T) {
+			update := pulumirpc.PropertyDiff_UPDATE
+			add := pulumirpc.PropertyDiff_ADD
+			delete := pulumirpc.PropertyDiff_DELETE
+			if forceNew {
+				update = pulumirpc.PropertyDiff_UPDATE_REPLACE
+				add = pulumirpc.PropertyDiff_ADD_REPLACE
+				delete = pulumirpc.PropertyDiff_DELETE_REPLACE
+			}
+
+			t.Run("unchanged", func(t *testing.T) {
+				runDetailedDiffTest(t, propertyMapElems("val1"), propertyMapElems("val1"), tfs, ps,
+					map[string]*pulumirpc.PropertyDiff{})
+			})
+
+			t.Run("changed non-empty", func(t *testing.T) {
+				runDetailedDiffTest(t,
+					propertyMapElems("val1"),
+					propertyMapElems("val2"), tfs, ps, map[string]*pulumirpc.PropertyDiff{
+						"foo[0].bar": {Kind: update},
```

Comment:
Hrm. This doesn't sound right. How did this work? Expected foo[0] to be updated, not foo[0].bar. THe "logical" diff was that we remove foo[0]=foo0 element and we add foo[0]=foo1 element that are different, not same.

Reply:
The CLI has no way of expressing that index 0 was both added and removed. This is the way we have to present the diff to the CLI to get around this.